src/HydraAuction/Delegate.hs

```diff
@@ -0,0 +1,107 @@
+module HydraAuction.Delegate (
+  delegateStep,
+  DelegateEvent (..),
+  DelegateInput (..),
+  DelegateRunnerT (..),
+  execDelegateRunnerT,
+) where
+
+-- Prelude imports
+import Prelude
+
+-- Haskell imports
+import Control.Monad.Except (MonadError (throwError))
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.State (MonadState, StateT (..), evalStateT, get, put)
+import Control.Monad.Trans (MonadTrans (lift))
+import Control.Monad.Trans.Except (runExceptT)
+
+-- HydraAuction imports
+
+import HydraAuction.Delegate.Interface (
+  DelegateError (..),
+  DelegateResponse (..),
+  FrontendKind (..),
+  FrontendRequest (..),
+  KnownFrontendRequest (..),
+ )
+import HydraAuction.Hydra.Interface (HydraEvent (..))
+import HydraAuction.Types (AuctionStage (..), AuctionTerms (..))
+
+-- FIXME: add client authentication (where is issue on this)
+
+data DelegateEvent
+  = Start
+  | AuctionStageStarted AuctionStage
+  | HydraEvent HydraEvent
+
+data DelegateInput
+  = DelegateEvent DelegateEvent
+  | FrontendRequest FrontendRequest
+data DelegateState = NoClient | HasClient FrontendKind AuctionTerms
```

Comment:
Why don't we just bundle the `AuctionTerms` in each request the client makes? This way the delegate is not tied to a specific auction, but can handle more than one. This would also simplify the issue with threads and state.

Reply:
> This way the delegate is not tied to a specific auction.

It is tied in scope of our milestones. And probably will stay so until Hydra will support partial commiting/decommiting or something like that.

For case of multiple auctions we could just add auction id handle to requests and this still will work.

Giving AuctionTerms every time complicates further implementation of checking of client authenticity, adds possibility of changing AuctionTerms for hacking or just providing strange errors (cuz delegate wont find UTxO by changed AuctionTerms).

Reply:
Yes, it waits for the client, and this is quite correct state, cuz delegate should exist before auction has started.